### PR TITLE
Added optimization baselines

### PR DESCRIPTION
**Pull request by elaude-pf**
### TL;DR

Added optimization capabilities to Constellaration with Augmented Lagrangian Method (ALM) and SciPy-based optimization runners.

### What changed?

- Added a new `very_low_fidelity` preset to VMEC settings for faster optimization runs
- Implemented Augmented Lagrangian optimization framework in a new `optimization` package
- Created optimization runners for both ALM and SciPy-based approaches
- Added launch scripts for different optimization problems (geometric, MHD stable, simple-to-build stellarators)
- Added utility modules for memoization, multiprocessing, and console silencing
- Added support for multithreaded VMEC runs via a new `num_threads_vmec` parameter
- Refined problem definitions and constraint handling

### How to test?

Run one of the launch scripts to test the optimization capabilities:

```bash
# For ALM optimization of a geometric problem
python -m constellaration.optimization.launch_alm_geometric_problem

# For SciPy optimization of a simple-to-build stellarator
python -m constellaration.optimization.launch_scipy_minimize_simple_to_build_stellarator

# For ALM optimization of MHD stable stellarator with different aspect ratios
python -m constellaration.optimization.launch_alm_mhd_stable_stellarator
```

### Why make this change?

This change enables efficient optimization of stellarator designs by:
1. Providing multiple optimization approaches (ALM and SciPy-based)
2. Supporting parallel evaluation of design candidates
3. Adding lower fidelity options for faster initial optimization
4. Implementing proper constraint handling for complex stellarator design problems

The optimization framework allows users to efficiently search the design space for stellarators that meet specific criteria while balancing competing objectives like aspect ratio, quasi-isodynamicity, and MHD stability.

### CODE REVIEW

File: src/constellaration/optimization/augmented_lagrangian.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+import dataclasses
+
+import jax.numpy as jnp
+import pydantic
+
+
+@dataclasses.dataclass
+class AugmentedLagrangianState:
```
Comment (mariap-pf) [inline src/constellaration/optimization/augmented_lagrangian.py:8]: can this also be a pydantic basemodel?

Comment (elaude-pf) [reply src/constellaration/optimization/augmented_lagrangian.py:8]: Yes

File: src/constellaration/problems_test.py
Comment (mariap-pf) [inline src/constellaration/problems_test.py:1]: this needs to be moved to tests/

File: src/constellaration/optimization/augmented_lagrangian.py (hunk 1 above)
Comment (mariap-pf) [reply src/constellaration/optimization/augmented_lagrangian.py:8]: yes, so you have to initialize the class as AugmentedLagrangianState(pydantic.BaseModel, arbitrary_types_allowed=True)

File: src/constellaration/optimization/augmented_lagrangian.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    """Decrement trust region bounds."""
+
+    penalty_parameters_max: float = 1e8
+    """Upper cutoff for penalty parameter."""
+
+    bounds_min: float = 0.05
+    """Lower cutoff for trust region bounds."""
+
+
+def augmented_lagrangian_function(
```
Comment (mariap-pf) [inline src/constellaration/optimization/augmented_lagrangian.py:34]: missing types

File: src/constellaration/optimization/augmented_lagrangian_runner.py [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+                        - 100 * metrics.vacuum_well,
+                    ]
+                )
+        else:
+            raise RuntimeError("Problem not supported")
+
+        return ((objective, constraints), metrics)
+
+
+def logging(k, n_function_evals, state: al.AugmentedLagrangianState):
```
Comment (mariap-pf) [inline src/constellaration/optimization/augmented_lagrangian_runner.py:124]: can this function be private for this module, and moved at the bottom?

Comment (mariap-pf) [reply src/constellaration/optimization/augmented_lagrangian_runner.py:124]: also there are missing types

File: src/constellaration/optimization/augmented_lagrangian_runner.py [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+) -> tuple[
+    tuple[jnp.ndarray, jnp.ndarray], forward_model.ConstellarationMetrics | None
+]:
+    with tempfile.TemporaryDirectory() as _:
+        boundary = unravel_and_unmask_fn(jnp.asarray(x * scale))
+
+        metrics = None
+        try:
+            # FIXME(amerlo): replace with better function.
+            with silence_console.silence_stderr():
```
Comment (mariap-pf) [inline src/constellaration/optimization/augmented_lagrangian_runner.py:43]: do we need to silence stderr?

File: src/constellaration/optimization/augmented_lagrangian_runner.py [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+    )
+    return surface.model_copy(
+        update=dict(
+            r_cos=fourier_coefficients_mask,
+            z_sin=fourier_coefficients_mask,
+        ),
+    )
+
+
+def main(
```
Comment (mariap-pf) [inline src/constellaration/optimization/augmented_lagrangian_runner.py:329]: could this be removed from here and each if branch moved to their respective launch script?

File: src/constellaration/optimization/scipy_minimize_runner.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+)
+
+NAN_TO_HIGH_VALUE = 10.0
+
+
+global n_function_evals
+n_function_evals = 0
+
+
+def _forward_model(
```
Comment (mariap-pf) [inline src/constellaration/optimization/scipy_minimize_runner.py:29]: private functions go after the public ones

File: src/constellaration/mhd/vmec_settings.py [hunk 7]
```diff
@@ -60,6 +60,7 @@ class VmecPresetSettings(pydantic.BaseModel):
         "high_fidelity",
         "low_fidelity",
         "ultra_low_fidelity",
```
Comment (amerlo-pf) [inline src/constellaration/mhd/vmec_settings.py:62]: OK in using `very_low_fidelity`, but then remove `ultra_low_fidelity`.

File: src/constellaration/optimization/scipy_minimize_runner.py [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+                    100 * problem._vacuum_well_lower_bound - 100 * metrics.vacuum_well,
+                ]
+            )
+    else:
+        raise RuntimeError("Problem not supported")
+
+    return constraints
+
+
+def logging(n_function_evals, objective, constraints):
```
Comment (mariap-pf) [inline src/constellaration/optimization/scipy_minimize_runner.py:150]: this is a copy from another module, can we move it in a shared util?

File: src/constellaration/optimization/settings.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    options: dict
+    method: str
+
+
+class OptimizationSettings(pydantic.BaseModel):
+    optimizer_settings: AugmentedLagrangianMethodSettings | ScipyMinimizeSettings
+    max_poloidal_mode: int
+    max_toroidal_mode: int
+    infinity_norm_spectrum_scaling: float
+    fidality: Literal[
```
Comment (mariap-pf) [inline src/constellaration/optimization/settings.py:34]: nit: fidelity

File: src/constellaration/utils/util.py
Comment (mariap-pf) [inline src/constellaration/utils/util.py:1]: can we rename to seed_util? util is too general

File: src/constellaration/optimization/augmented_lagrangian.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+def update_augmented_lagrangian_state(
+    x: jnp.ndarray,
+    objective: jnp.ndarray,
+    constraints: jnp.ndarray,
+    state: AugmentedLagrangianState,
+    settings: AugmentedLagrangianSettings,
+    penalty_parameters: jnp.ndarray | None = None,
+    bounds: jnp.ndarray | None = None,
+) -> AugmentedLagrangianState:
```
Comment (amerlo-pf) [inline src/constellaration/optimization/augmented_lagrangian.py:60]: Maybe a docstring here

File: src/constellaration/optimization/augmented_lagrangian_runner.py [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+
+    print(k, optimizer_stats)
+
+
+def run(
+    boundary: rz_fourier.SurfaceRZFourier,
+    settings: optimization_settings.OptimizationSettings,
+    problem: problems.MHDStableQIStellarator | problems.SingleObjectiveProblem,
+    aspect_ratio_upper_bound: float | None = None,
+):
```
Comment (amerlo-pf) [inline src/constellaration/optimization/augmented_lagrangian_runner.py:140]: Maybe a docstring here with return values

File: src/constellaration/optimization/augmented_lagrangian_runner.py [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+            jnp.minimum(
+                settings.optimizer_settings.oracle_settings.budget_max,
+                budget + settings.optimizer_settings.oracle_settings.budget_increment,
+            )
+        )
+
+        logging(k + 1, n_function_evals, state)
+
+
+def _build_surface_rz_fourier_mask(
```
Comment (amerlo-pf) [inline src/constellaration/optimization/augmented_lagrangian_runner.py:309]: please use the function under rz_surface_fourier

File: src/constellaration/optimization/scipy_minimize_runner.py [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+    unravel_and_unmask_fn: Callable[[jnp.ndarray], rz_fourier.SurfaceRZFourier],
+    settings: (forward_model.ConstellarationSettings),
+) -> forward_model.ConstellarationMetrics | None:
+    with tempfile.TemporaryDirectory() as _:
+        boundary = unravel_and_unmask_fn(jnp.asarray(x * scale))
+
+        metrics = None
+        try:
+            # FIXME(amerlo): replace with better function.
+            with silence_console.silence_stderr():
```
Comment (amerlo-pf) [inline src/constellaration/optimization/scipy_minimize_runner.py:41]: Please remove this, we do not strictly needed it

File: src/constellaration/optimization/scipy_minimize_runner.py [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+        settings_kwargs.update(
+            dict(
+                qi_settings=None,
+                turbulent_settings=None,
+            )
+        )
+
+    forward_model_settings = forward_model.ConstellarationSettings(**settings_kwargs)
+
+    fm = memoize.Memoize(
```
Comment (amerlo-pf) [inline src/constellaration/optimization/scipy_minimize_runner.py:204]: I would suggest not to use memoize, I would not use it and also remove the module

File: src/constellaration/optimization/scipy_minimize_runner.py [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+        ),
+        x0=x0,
+        method=settings.optimizer_settings.method,
+        options=settings.optimizer_settings.options,
+        constraints=constraints,
+        callback=callback,
+    )
+
+
+def _build_surface_rz_fourier_mask(
```
Comment (amerlo-pf) [inline src/constellaration/optimization/scipy_minimize_runner.py:269]: please use the mask function in the rz_fourier_surface module

File: src/constellaration/optimization/scipy_minimize_runner.py [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+    )
+    return surface.model_copy(
+        update=dict(
+            r_cos=fourier_coefficients_mask,
+            z_sin=fourier_coefficients_mask,
+        ),
+    )
+
+
+def main(
```
Comment (amerlo-pf) [inline src/constellaration/optimization/scipy_minimize_runner.py:289]: same comment as above, let us remove it from here and only keep it in the launcher script

File: src/constellaration/optimization/settings.py [hunk 17]
```diff
@@ -0,0 +1,42 @@
+from typing import Literal
+
+import pydantic
+
+import constellaration.optimization.augmented_lagrangian as al
+
+
+class NevergradSettings(pydantic.BaseModel):
```
Comment (amerlo-pf) [inline src/constellaration/optimization/settings.py:8]: these might deserve a docstring for each parameter

File: src/constellaration/utils/multiprocessing_context.py [hunk 18]
```diff
@@ -0,0 +1,16 @@
+import multiprocessing
```
Comment (amerlo-pf) [inline src/constellaration/utils/multiprocessing_context.py:1]: GENERAL: I would remove this and use instead `multiprocessing_context = multiprocessing.get_context("forkserver")`

File: src/constellaration/utils/util.py
Comment (amerlo-pf) [reply src/constellaration/utils/util.py:1]: I would agree

**Comment by mariap-pf** [comment]
I think all the example scripts should actually be outside of src/.., because otherwise they will be packaged and it makes no sense to include them in the package. So please create a top level dir `examples` or similar.

File: src/constellaration/optimization/scipy_minimize_runner.py (hunk 8 above)
Comment (elaude-pf) [reply src/constellaration/optimization/scipy_minimize_runner.py:150]: it is slightly different as it does not have k and state object

File: src/constellaration/problems_test.py
Comment (elaude-pf) [reply src/constellaration/problems_test.py:1]: Indeed. it already exists there

**Comment by scadena-pf** [comment]
@elaude-pf How about creating at the top level a folder called optimization_examples/

**Review by mariap-pf — CHANGES_REQUESTED** [review]
Thanks for doing all the changes! with the example files, I meant we should add a top level dir (like src) with all the scripts, not a a sub dir of optimization!
so we would have:
CONSTELLARATION/
├──examples/   (or optimization_examples)
├── notebooks/
├── src/
│   └── constellaration/
└── tests/

**Review by mariap-pf — APPROVED** [review]
Thanks!!! Please also fix the error from CI :D
